plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -928,8 +928,8 @@ private void finishOptimize(ConnectorSession session, IcebergTableExecuteHandle
             newFiles.add(builder.build());
         }
 
-        if (scannedFiles.isEmpty() && newFiles.isEmpty()) {
-            // Table scan turned out to be empty, nothing to commit
+        // Table scan turned out to be empty, or the procedure doesn't reduce file counts, nothing to commit
+        if ((scannedFiles.isEmpty() && newFiles.isEmpty()) || scannedFiles.size() <= newFiles.size()) {
```

Comment:
This is an interesting case. It may be an improvement, but not necessarily. 
Under what circumstances could we produce more files than originally?

this can happen for example when optimize is invoked without file size threshold (and so picks up big files), and the connector is configured with target file size  to create files no bigger than X.
in such case, proceeding with a rewrite seems appropriate: new files better align with desired sizes than the original ones

other case where this may happen is when the table partitioning changed and new files are partitioned more granularity. again, this adheres better to the table configuration than the state before the change

Reply:
Other cases where i would want to proceed with optimize even if it has more new files are:

- application of delete deltas
- sorting files (once we support file-level sorting)

Reply:
Then, can we close this PR and also #10785?

Reply:
@ebyhr the current logic seems too simplistic. The referenced issue is an enhancements, which is generally desirable ("don't do unnecessary work"). There may be other ways of achieving that goal, that feel safer.

Reply:
Let me close only this PR.